src/main/kotlin/pt/up/fe/ni/website/backend/service/RoleService.kt

```diff
@@ -1,46 +1,113 @@
-import pt.up.fe.ni.website.backend.model.Activity
+import pt.up.fe.ni.website.backend.dto.entity.RoleDto
 import pt.up.fe.ni.website.backend.model.PerActivityRole
 import pt.up.fe.ni.website.backend.model.Role
-import pt.up.fe.ni.website.backend.model.permissions.Permission
 import pt.up.fe.ni.website.backend.model.permissions.Permissions
+import pt.up.fe.ni.website.backend.repository.GenerationRepository
 import pt.up.fe.ni.website.backend.repository.PerActivityRoleRepository
 import pt.up.fe.ni.website.backend.repository.RoleRepository
+import pt.up.fe.ni.website.backend.service.activity.ActivityService
 
 @Service
+@Transactional
 class RoleService(
     private val roleRepository: RoleRepository,
-    private val perActivityRoleRepository: PerActivityRoleRepository
+    private val perActivityRoleRepository: PerActivityRoleRepository,
+    private val generationRepository: GenerationRepository,
+    private val accountService: AccountService,
+    private val activityService: ActivityService
 ) {
 
-    fun grantPermissionToRole(role: Role, permission: Permission) {
-        role.permissions.add(permission)
+    fun getRole(roleId: Long): Role {
+        val role = roleRepository.findById(roleId).orElseThrow {
+            throw NoSuchElementException(ErrorMessages.roleNotFound(roleId))
+        }
+        return role
+    }
+    fun getAllRoles(): List<Role> = roleRepository.findAll().toList()
+    fun grantPermissionToRole(roleId: Long, permissions: Permissions) {
+        val role = getRole(roleId)
+        role.permissions.addAll(permissions)
         roleRepository.save(role)
     }
 
-    fun revokePermissionFromRole(role: Role, permission: Permission) {
-        role.permissions.remove(permission)
+    fun revokePermissionFromRole(roleId: Long, permissions: Permissions) {
+        val role = getRole(roleId)
+        role.permissions.removeAll(permissions)
         roleRepository.save(role)
     }
 
-    fun grantPermissionToRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
-        val foundActivity = activity.associatedRoles
+    fun grantPermissionToRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
+        val foundPerActivityRole = activity.associatedRoles
             .find { it.activity == activity } ?: PerActivityRole(Permissions())
-        foundActivity.role = role
-        foundActivity.activity = activity
+        foundPerActivityRole.role = role
+        foundPerActivityRole.activity = activity
 
-        foundActivity.role = role
-        foundActivity.permissions.add(permission)
-        perActivityRoleRepository.save(foundActivity)
+        foundPerActivityRole.permissions.addAll(permissions)
+        perActivityRoleRepository.save(foundPerActivityRole)
+        if (activity.associatedRoles.find { it.activity == activity } == null) {
```

Comment:
This check is already performed above. I'd suggest extracting it to a variable and use it on line 47 and 53!

Reply:
I can now remove this logic because it was due to ManyToOne transaction shenanigans that can be fixed with the new function `TestUtils.startNewTransaction()`

Reply:
Nice!